cpp/ql/src/Telemetry/Diagnostics.qll

```diff
@@ -0,0 +1,29 @@
+import cpp
+
+/**
+ * A syntax error.
+ */
+class SyntaxError extends CompilerError {
+  SyntaxError() {
+    this.getTag().matches("exp_%") or
+    this.getTag() =
+      [
+        "bad_data_member_initialization", "bad_pure_specifier", "bad_return", "bad_uuid_string",
+        "literal_without_initializer", "missing_class_definition", "missing_exception_declaration",
+        "nonstd_const_member_decl_not_allowed", "operator_name_not_allowed",
+        "wide_string_invalid_in_asm"
+      ]
+  }
+}
+
+/**
+ * A cannot open file error.
+ * Typically this is due to a missing include.
+ */
+class CannotOpenFile extends CompilerError {
+  CannotOpenFile() { this.hasTag(["cannot_open_file", "cannot_open_file_reason"]) }
+
+  string getIncludedFile() {
+    result = this.getMessage().regexpCapture("cannot open source file '([^']+)'", 1)
```

Comment:
Note that this regex does not what you want in the case of `cannot_open_file_reason`.

Reply:
Yea, I've no real way of testing this as I've never seen the extractor produce this error message. I suppose a `.*` at the end of the regex might be reasonable?

Reply:
Are the file names always quoted? This is non-obvious from the frontend source code.

Reply:
Yes, I could test this on some real databases. But an integration test is clearly needed then. Qltest cannot test this situation due to the absence of `SEMMLE_CPP_MISSING_INCLUDES_NOT_FATAL`.

Reply:
I think that having some integration test for this is the only thing that still needs addressing?

Reply:
Here it is: https://github.com/github/semmle-code/pull/51788